### ISSUE TITLE
Java 17 records don't work in native image with GraalVM 21.3.0 JDK 17

### ISSUE THREAD

**Issue by michael-simons** [quarkusio/quarkus#20891]
### Describe the bug

Java 17 records don't work with GraalVM 21.3.0 when used as return types of resources that produce `application/json`. They also fail with SmallRye GraphQL. 

In the reproducer I used `quarkus-resteasy-jackson` because Jackson supports records out of the box without any configuration.

When I create a native image like that with GraalVM 21.3 it works as expected.

```java
public class App {

	public static void main(String[] args) {

		System.out.println(new F("x"));
	}

	static record F(String x) {
	}
}
```

### Expected behavior

Quarkus native should produce a binary.

### Actual behavior

Native compilation fails with:

```
Error: com.oracle.svm.hosted.substitute.DeletedElementException: Unsupported method java.lang.Class.getConstantPool() is reachable: The declaring class of this element has been substituted, but this element is not present in the substitution class
To diagnose the issue, you can add the option --report-unsupported-elements-at-runtime. The unsupported element is then reported at run time when it is accessed the first time.
Detailed message:
Trace: 
	at parsing java.lang.System$2.getConstantPool(System.java:2262)
Call path from entry point to java.lang.System$2.getConstantPool(Class): 
	at java.lang.System$2.getConstantPool(System.java:2262)
	at java.lang.reflect.RecordComponent.declaredAnnotations(RecordComponent.java:198)
	at java.lang.reflect.RecordComponent.getAnnotation(RecordComponent.java:181)
	at io.smallrye.config.ConfigMappingInterface.getPropertyName(ConfigMappingInterface.java:691)
	at io.smallrye.config.ConfigMappingInterface.getPropertyDef(ConfigMappingInterface.java:622)
	at io.smallrye.config.ConfigMappingInterface.getProperties(ConfigMappingInterface.java:607)
	at io.smallrye.config.ConfigMappingInterface.createConfigurationInterface(ConfigMappingInterface.java:564)
	at io.smallrye.config.ConfigMappingInterface.access$000(ConfigMappingInterface.java:27)
	at io.smallrye.config.ConfigMappingInterface$1.computeValue(ConfigMappingInterface.java:32)
	at io.smallrye.config.ConfigMappingInterface$1.computeValue(ConfigMappingInterface.java:30)
	at com.oracle.svm.core.jdk.Target_java_lang_ClassValue.get(JavaLangSubstitutions.java:596)
	at io.quarkus.runtime.configuration.Substitutions$Target_ConfigMappingInterface.getConfigurationInterface(Substitutions.java:86)
	at io.smallrye.config.ConfigMappingContext.lambda$getKeyConverter$6(ConfigMappingContext.java:142)
	at io.smallrye.config.ConfigMappingContext$$Lambda$1290/0x00000007c1585240.apply(Unknown Source)
	at sun.security.ec.ParametersMap$1.get(ParametersMap.java:78)
	at com.oracle.svm.core.jdk.SystemPropertiesSupport.initializeLazyValue(SystemPropertiesSupport.java:216)
	at com.oracle.svm.core.jdk.SystemPropertiesSupport.getProperty(SystemPropertiesSupport.java:169)
	at com.oracle.svm.core.jdk.Target_java_lang_System.getProperty(JavaLangSubstitutions.java:287)
	at com.oracle.svm.jni.JNIJavaCallWrappers.jniInvoke_ARRAY_System_getProperty_deeeaa72a006d330408a3b7d002c7533e108bc28(generated:0)
```

### How to Reproduce?

Given the following dependency:

```xml
<dependency>
	<groupId>io.quarkus</groupId>
	<artifactId>quarkus-resteasy-jackson</artifactId>
</dependency>
```

This `record`:

```java
public record Thing(String name) {
}
```

and the following resource

```java
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

@Path("/hello")
public class ExampleResource {

	@GET
	@Produces(MediaType.APPLICATION_JSON)
	public Thing hello() {
		return new Thing("Hello RESTEasy");
	}
}
```

A request like `curl localhost:8080/hello` returns 

```json
{"name":"Hello RESTEasy"}
```

The whole project is here: [restrecords.zip](https://github.com/quarkusio/quarkus/files/7379169/restrecords.zip)

Reproduce with

```
 ./mvnw clean package -Pnative
```

on GraalVM 21.3 Java 17

An example using SmallRye GraphQL is available here: https://github.com/michael-simons/neo4j-aura-quarkus-graphql/tree/jdk17

### Output of `uname -a` or `ver`

Darwin bs62.XXX 20.6.0 Darwin Kernel Version 20.6.0: Mon Aug 30 06:12:21 PDT 2021; root:xnu-7195.141.6~3/RELEASE_X86_64 x86_64

### Output of `java -version`

openjdk version "17.0.1" 2021-10-19 OpenJDK Runtime Environment GraalVM CE 21.3.0 (build 17.0.1+12-jvmci-21.3-b05) OpenJDK 64-Bit Server VM GraalVM CE 21.3.0 (build 17.0.1+12-jvmci-21.3-b05, mixed mode, sharing)

### GraalVM version (if different from Java)

21.3.0

### Quarkus version or git rev

2.3.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d)

**Comment by geoand** [comment]
@zakkak this is still an issue with the released version of `21.3`. Do you know anything about this?

**Comment by zakkak** [comment]
@geoand not much.

After looking a bit into it, it appears to be an issue with smallrye-config not being _native-image friendly_ and not with Java records support in GraalVM or Quarkus.

GraalVM [doesn't provide `java.lang.Class.getConstantPool()`](https://github.com/oracle/graal/blob/9c3e90756b204d1cfb107acbb62501a860468218/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/hub/DynamicHub.java#L1646-L1647), yet smallrye-config somehow ends up invoking it judging by the [the stacktrace](https://github.com/quarkusio/quarkus/issues/20891#issue-1031090789).

Two options I see are:
1. Find why smallrye-config ends up following that path for a record, is this by mistake?
2. Susbstitute one of the methods in the call-tree to avoid the issue (probably by stopping early when working on a record?).

HTH

**Comment by geoand** [comment]
@radcortez ^

**Comment by radcortez** [comment]
I'll have a look.

**Comment by zakkak** [comment]
Upstream issue: https://github.com/oracle/graal/issues/3984

**Comment by crilofer** [comment]
Same issue here. Working adding `--report-unsupported-elements-at-runtime` but shouldn't be the best option IMO...

**Comment by Aleksandr-Filichkin** [comment]
any progress so far?

**Comment by jorsol** [comment]
> any progress so far?

haven't tested myself, but it looks like Mandrel 21.3.1.1 fixes this issue:
https://github.com/graalvm/mandrel/releases/tag/mandrel-21.3.1.1-Final

**Comment by jerboaa** [comment]
Upstream Graal VM will have this fixed in 21.3.2 and 22.1 releases (April 2022). Mandrel included this fix in `[21.3.1.1-Final](https://github.com/graalvm/mandrel/releases/tag/mandrel-21.3.1.1-Final)` early, but will sync with upstream in April.

**Comment by juanmbellini** [comment]
In my case I have the issue when instantiating a record. If record is declared, but the instantiation is not called, then native-image works.

**Comment by pjgg** [comment]
```java
package org.acme;

import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

@Path("/hello")
public class GreetingResource {

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        System.out.printf(new Thing("asd").toString());
        return "Hello RESTEasy";
    }
}
```

Based on the above reproducer given by @michael-simons  I could reproduce it with `ubi-quarkus-mandrel:21.3.0.0-Final-java17` 

Fails: `mvn clean package -Pnative -Dquarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-mandrel:21.3.0.0-Final-java17 -Dquarkus.platform.version=2.7.5.Final`

Is fixed if you use `ubi-quarkus-mandrel:21.3.1.1-Final-java17` as a build base image

Works: `mvn clean package -Pnative -Dquarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-mandrel:21.3.1.1-Final-java17 -Dquarkus.platform.version=2.7.5.Final`

Or directly you can use `ubi-quarkus-mandrel:21.3-java17`

**Comment by geoand** [comment]
Closing this as Quarkus `2.10` will use GraalVM 22.1.
